game/audio.cc

```diff
@@ -62,173 +67,191 @@ namespace {
+	// suppress center channel vocals
+	if(suppressCenterChannel && !m_preview) {
+		float diffLR;
+		for (size_t i=0; i<mixbuf.size(); i+=2) {
+			diffLR = begin[i] - begin[i+1];
+			begin[i] = diffLR;
+			begin[i+1] = diffLR;
+		}
 	}
-	void trackPitchBend(std::string const& name, double pitchFactor) {
-		auto it = tracks.find(name);
-		if (it == tracks.end()) return;
-		it->second->pitchFactor = pitchFactor;
+	return !eof;
+}
+
+double Music::duration() const {
+	double dur = 0.0;
+	for (auto& kv: tracks) dur = std::max(dur, kv.second->mpeg.audioQueue.duration());
+	return dur;
+}
+
+bool Music::prepare() {
+	bool ready = true;
+	for (auto& kv: tracks) {
+		FFmpeg& mpeg = kv.second->mpeg;
+		if (mpeg.terminating()) continue;  // Song loading failed or other error, won't ever get ready
+		if (mpeg.audioQueue.prepare(m_pos)) {
+			if (kv.first == "background" && m_preview && m_pos > 0) {
+				fvec_t* previewSamples = mpeg.audioQueue.makePreviewBuffer();
+				fvec_t* previewBeats = ScreenSongs::previewBeatsBuffer.get();
+				intptr_t readptr = 0;
+				fvec_t* tempoSamplePtr = new_fvec(Audio::aubio_hop_size);
+				std::lock_guard<std::recursive_mutex> l(Music::aubio_mutex);
+// 				std::clog << "audio/debug: file: " << mpeg.getFilename().filename().string() << std::endl;
+				Game* gm = Game::getSingletonPtr();
+				ScreenSongs* sSongs = static_cast<ScreenSongs *>(gm->getScreen("Songs"));
+				double pstart = sSongs->getSongs().currentPtr()->preview_start;
+				pstart = (std::isnan(pstart) ? 0.0 : pstart);
+				double first_period, first_beat;
+				std::vector<double> extra_beats;
+				Song::Beats& beats = sSongs->getSongs().currentPtr()->beats;
+				if (!sSongs->getSongs().currentPtr()->hasControllers()) {
+				while (readptr < previewSamples->length) {
+// 						std::clog << "audio/debug: readptr: " << readptr << std::endl;
+					tempoSamplePtr->data = &previewSamples->data[readptr];
+					aubio_tempo_do(aubioTempo.get(),tempoSamplePtr,previewBeats);
+					if (previewBeats->data[0] != 0) {
+						double beatSecs = aubio_tempo_get_last_s(aubioTempo.get());
+							if (beats.empty()) { // Store time and period of first detected beat.
+								first_beat = beatSecs;
+								first_period = aubio_tempo_get_period_s(aubioTempo.get());
+							}
+						beats.push_back(beatSecs + pstart);
+//						std::clog << "aubio/debug: beat at: " << aubio_tempo_get_last_ms(aubioTempo.get()) << "ms (" << aubio_tempo_get_last_s(aubioTempo.get()) << "s). frame: " <<  aubio_tempo_get_last(aubioTempo.get()) << ", period: " << aubio_tempo_get_period_s(aubioTempo.get()) << ", BPM: " << aubio_tempo_get_bpm(aubioTempo.get()) << ", confidence: " << double(aubio_tempo_get_confidence(aubioTempo.get())) << std::endl;
+					}
+					readptr += Audio::aubio_hop_size;
+				}
+					if (!beats.empty()) {
+						double newBeat = first_beat - first_period;
+						while (newBeat > 0.0) {
+// 							std::clog << "aubio/debug: Extrapolated a beat at: " << newBeat << "s." << std::endl;
```

Comment:
Because it's potentially useful information to have if we want to modify this in the future but leaving it in adds some not-so-minor delay and spam in the log.

Actually I've been thinking about this for a while but maybe we ought to add more log levels and reorganize how some of the existing messages are categorized.